### PR TITLE
feat(dedupe): per-table id_field/require_id overrides

### LINKED ISSUE #222: Per-table dedupe id_field (+ fix cross-table dedupe keyspace collision)
**Area:** dedupe — gap + latent correctness bug · found via WaveHouse-Stats dogfooding

**Expected:** each table dedupes on its own natural key, in its own keyspace.

**Actual:** two problems with the single global dedupe key:
1. **One global `id_field`** name applies deployment-wide (`internal/config/config.go:105-108`) — every table that wants dedupe must carry an identically-named column.
2. **Shared keyspace (latent bug):** `CheckAndMark` keys Pebble on the raw id value with **no table namespace** (`internal/dedupe/embedded.go:29`, `key := []byte(eventID)`). Equal id *values* across two different tables collide — the second is silently dropped as a "duplicate." Only bites when ids aren't globally unique across tables (e.g. per-table sequential ids), but it's silent data loss when it does.

**Impact (Stats):** we standardize on globally-unique `event_id`, so the collision doesn't bite us today — but it's a sharp edge for any multi-table deployment, and the single id_field forces a naming convention on every future table.

**Scope:** namespace the dedupe key by table (fixes #2 regardless), and allow a per-table id_field (fixes #1). The keyspace fix is the more urgent half.

Related: #219, #220, #221.

---
_From `WAVEHOUSE-FEEDBACK.md` (WaveHouse-Stats dogfooding); validated by code-read against `0f8826c` on 2026-06-04._

### LINKED ISSUE #370: bug(ingest): explicit null id_field slips past require_id and dedupes every row onto "<nil>"
**Area:** ingest · api — bug (silent data loss / correctness) · found via PR review (#219 follow-up)

**Expected:** With dedupe on, a row whose `id_field` is present but `null` (`{"event_id": null}`) is treated like a missing id — counted by `wavehouse_ingest_dedupe_missing_id_total` and rejected under `require_id: true` (else published un-deduped).

**Actual:** `internal/api/ingest.go` `processRecord` does `idVal, ok := data[h.IDField]`; an explicit JSON `null` decodes to a present key, so `ok == true` and the record skips the missing-id branch. `fmt.Sprint(nil)` yields the literal `"<nil>"`, and the dedup key is un-namespaced (`internal/dedupe/embedded.go:29`, `key := []byte(eventID)` — no table/scope prefix), so **every** null-id row across **every** table collides on the single key `"<nil>"`: the first is ingested, the rest are dropped as duplicates. `require_id` never fires and the metric never counts them.

**Impact:** Silent data loss whenever dedupe is on and a producer emits `null` for the id (common when a serializer writes null for an unset field). The `require_id` tripwire — whose whole point is guaranteeing a usable id — is defeated, and the omission is invisible in metrics.

Scope: treat a nil value as missing (`if !ok || idVal == nil`) so null flows through the existing missing-id path; add tests for `{"event_id": null}` in both `require_id` modes. Decide in the same fix whether empty-string `""` should also count as missing (same collision class). Separate latent issue: `fmt.Sprint` keys collide across types (`123` vs `"123"`).

Related: #219 (the missing-id handling this extends)

### PR DESCRIPTION
**Pull request by taitelee**
## Summary

Dedupe's `id_field` and `require_id` were single, deployment-wide settings. This PR makes them per-table and fixes two dedupe keying bugs found along the way:

- **Per-table overrides** (closes #222): a new optional `dedupe.tables` map overrides `id_field`/`require_id` per table; the flat `dedupe.*` settings remain the defaults for any table not listed. Setting a table's `id_field` to `""` disables dedupe for that table.
- **Table-namespaced keys** (part of #222): dedupe keys are now `table\x00id`, so the same id *value* in two different tables no longer collides. **Breaking:** the on-disk key format changes, so existing dedupe state is treated as new for one round after upgrade (best-effort idempotency; no stored data lost).
- **Explicit-null id fix** (closes #370): a present-but-null `id_field` (`{"event_id": null}`) was deduping every null row onto the literal `"<nil>"` key. A null id is now treated as missing — logged, counted, and rejected under `require_id` (or published un-deduped otherwise).

## Related Issues

Closes #222
Closes #370

### CODE REVIEW

File: internal/api/ingest_test.go [hunk 1]
```diff
@@ -530,6 +530,124 @@ func TestIngest_Dedup_RequireID_Rejects(t *testing.T) {
...
+	h.DedupeTables = map[string]DedupeOverride{
+		"clicks": {RequireID: ptr(true)}, // this table is strict
+	}
+
+	w := httptest.NewRecorder()
+	h.Handle(w, ingestRequest(t, "clicks", map[string]any{"page": "/home"})) // no event_id
+	assert.Equal(t, http.StatusBadRequest, w.Code, "clicks overrides require_id=true, so a missing id is rejected")
+	assert.Contains(t, w.Body.String(), "missing dedupe id field")
+	assert.Nil(t, pub.LastMessage())
+}
```
[the remark this answers is not in the document]
Comment (taitelee) [reply internal/api/ingest_test.go:598]: Added `testutil.AssertJSONErrorResponse(t, w)` so this strict-mode rejection path asserts the JSON error envelope like its sibling `TestIngest_Dedup_NullIDField_RequireID_Rejects`.

File: docs/src/content/docs/api.md [hunk 2]
```diff
@@ -225,7 +225,7 @@ The body is a **flat JSON object** whose keys must match column names in the tar
...
-| 400 | `{"error":"missing dedupe id field \"event_id\""}` | Only when dedupe is enabled with `dedupe.require_id: true` and the row lacks the configured `id_field`. With `require_id: false` (the default) the row is instead published un-deduped. Either way — reject or publish — the row is logged at `WARN` and counted by `wavehouse_ingest_dedupe_missing_id_total`. In a batch this is a per-record failure, not a whole-request error. |
+| 400 | `{"error":"missing dedupe id field \"event_id\""}` | Only when dedupe is enabled with `dedupe.require_id: true` and the row lacks the configured `id_field` — whether the field is missing or sent as an explicit JSON `null`. With `require_id: false` (the default) the row is instead published un-deduped. Either way — reject or publish — the row is logged at `WARN` and counted by `wavehouse_ingest_dedupe_missing_id_total`. In a batch this is a per-record failure, not a whole-request error. |
```
[the remark this answers is not in the document]
Comment (taitelee) [reply docs/src/content/docs/api.md]: Updated.

**Comment by EricAndrechek** [comment]
So just to be sure I'm understanding this correctly:
- you can now NOT set everything via ENV variables – tables dedupe settings become possible ONLY via the config file now?
- is there a way to default to disabled and turn on only specific tables, or only the inverse via setting a table's `id_field` to `""` to disable dedupe for it?

**Comment by taitelee** [comment]
To answer Eric's questions:

- Yes. The tables dedupe settings become possible only via the config file. 
- Both. Verifying this exposed a bug. The flat `dedupe.*` settings are the defaults, and a tables entry overrides them. An empty `id_field` means "don't dedupe". Turning one table off (`id_field: ""` under tables:) already worked. The opposite — off by defauopt-in was an undocumented env var trick.

GOing to fix now. Leaving` id_field` out still defaults to `event_id`. An explicit `""` (in the file, or `WH_DEDUPE_ID_FIELD=`) now sticks and turns dedupe off by default:

We realized that having dedupe settings requiring a config file could get messy for the user, and solely using env variables for dedupe settings can be a complex task, so we are deferring this until we have solved issue #48.
